### DISCUSSION TITLE
Generic argument in TestCaseSource

### DISCUSSION THREAD

**Discussion by Maesla** [nunit/nunit#4291] [Q&A] [answered]
Hello,

Is it possible to have a generic argument in a TestCaseSource?

I am having this exception:
`System.InvalidOperationException : Late bound operations cannot be performed on types or methods for which ContainsGenericParameters is true.`

Dummy example:
```csharp
using NUnit.Framework;
using System.Collections.Generic;

public class Dummy
{
    [Test, TestCaseSource(typeof(Dummy), nameof(DummyGet), new object[] { "" })]
    public void DummyFooTest(Foo foo)
    {
        Assert.That(foo.id, Is.EqualTo(1));
    }

    [Test, TestCaseSource(typeof(Dummy), nameof(DummyGet), new object[] { "" })]
    public void DummyBarTest(Bar bar)
    {
        Assert.That(bar.id, Is.EqualTo(2));
    }

    private static IEnumerable<T> DummyGet<T>(string dummyParameter)
    {
        yield return default(T);
        yield return default(T);
        yield return default(T);
    }

    public class Foo
    {
        public int id;

        public Foo()
        {
            id = 1;
        }
    }

    public class Bar
    {
        public int id;

        public Bar()
        {
            id = 2;
        }
    }
}

```

Thank you

**Comment by mikkelbu** [comment]
I don't think it is possible. Perhaps one can do something similar using a generic TestFixture
```csharp
    [TestFixture(typeof(Foo))]
    [TestFixture(typeof(Bar))]
    public class Dummy<T>
...
```

But please note that `default(T)` when `T` is a class will be `null` and not call the default constructor. You can do this using `new T()`, if you have a type constraint saying that `T` has a default constructor - i.e. `where T: new()`.

**Reply by Maesla** [reply]
The DummyGet<T> method is just an example. The real one doesn't even have a default expression.

The key is how you can pass the actual type to the generic method using TestCaseSource

I am currently using a workarround, defining a non generic method that calls the generic method, but I wanted to know if I could call directly to the generic method without passing through the middle method

Thank you

**Reply by mikkelbu** [reply]
I would have suggested your workaround as the best solution

**Comment by manfred-brands** [comment, marked as answer]
There is currently no such way to pass that extra type parameter.

**Reply by Maesla** [reply]
Ok, thank you
